feat(solr): add exact-match field and ranking results

PouyaMohseni opened the pull request:
- In #407, the ranking and fuzzy matching of search results were discussed. The system previously used n-gram matching to compare the query with language labels.
- This adds exact matching which, in contrast to n-gram matching, requires an exact match of the term.
- In the final results, exact matches are weighted twice as strongly as n-gram matches.

PouyaMohseni committed:
feat(solr): add exact-match field and ranking results

PouyaMohseni committed:
feat(solr): add exact-match field and ranking results

dchiller reviewed (commented):
Can you say more about why we are adding this? What problem does this solve?

I see in the linked issue that both fuzzy matching and ranking are discussed. It looks to me like this is more related to the ranking than the fuzzy matching. Is that right?

What is the type of ranking that we are hoping to see? And how does this approach get us towards that ranking? 

For example, I see in the issue that we want a search of `tar` to return more the instrument with the name "tar" over the ones with the name "guitar". I imagine your solution will achieve this result. But what if I search "ta"? Do I still want "tar" to come up before "guitar" in the results? Will this achieve that? Would something like Solr's Edge N-gram Tokenizer be more what we want? 

I'm not sure that that's exactly what we want, but it certainly seems to me like a case where "tar" returns "tar" over "guitar", but "ta" doesn't isn't necessarily what we want.

PouyaMohseni committed:
fix: enhance Solr exact match

- remove unused fieldType
- add and weight text_exact in \select compared to text_ngram
- moved wikidata_id_s to text_exact from text_ngram

PouyaMohseni committed:
fix: enhance Solr exact match

- remove unused fieldType
- add and weight text_exact in \select compared to text_ngram
- moved wikidata_id_s to text_exact from text_ngram

dchiller commented on solr/cores/conf/schema.xml:
```diff
@@ -79,16 +79,33 @@
     <field name="text" type="text_general" indexed="true" stored="true" multiValued="true"/>
 
+    <!-- exact match field -->
+    <field name="text_exact" type="text_vector" indexed="true" stored="true" multiValued="true"/>
```
> I think both text (text_ngram) and text_exact should not be stored but are multiValued.

Agreed about not storing them. Not sure about `multiValued` though... what are the multiple values?

As far as I can see, both are creating by coping a number of other fields in to this field and indexing the result...

ahankinson replied:
I believe text fields, particularly those that are used as targets for copyfields, should generally be multivalued fields. As fields are copied into the text field I believe they are kept as a distinct value in that field, and not simply appended as one big string. 

See: https://solr.apache.org/guide/solr/latest/indexing-guide/copy-fields.html

"In the example above, if the text destination field has data of its own in the input documents, the contents of the cat field will be added as additional values – just as if all of the values had originally been specified by the client. Remember to configure your fields as `multivalued="true"` if they will ultimately get multiple values (either from a multivalued source or from multiple copyField directives)."

ahankinson commented on solr/cores/conf/solrconfig.xml:
```diff
@@ -71,10 +71,13 @@
             <int name="rows">10</int>
-            <str name="df">text</str>
+            <str name="defType">edismax</str>
+            <str name="df">text_exact text_ngram</str>
```
If you specify `qf` you don't need to also specify `df`. 

This should probably look more like this:

```xml
 <arr name="qf">
    <str>text_exact^2</str>
    <str>text_ngram</str>
</arr>
```

PouyaMohseni committed:
fix: update qf configuration and remove redundant df setting

PouyaMohseni committed:
fix: update qf configuration and remove redundant df setting
